A11Y fixes for toolbar and some typo fixes for aria-labelledbyAccessibility

akash07k opened the pull request:
There were some typos in aria-labelled by.
Fixed them.
Previously they were written as aria-labeledby which I noticed that is wrong.
Also, made some improvements in the toolbar accessibility.
Excluded the assets/.cache directory from the VCS.

jtojnar commented on assets/js/selfoss-ui.js:
```diff
@@ -102,11 +102,13 @@ selfoss.ui = {
-                    <button id="nav-logout" title={selfoss.ui._('logoutbutton')} aria-label={selfoss.ui._('logoutbutton')} accesskey="l"></button>
-                    <button id="nav-login" title={selfoss.ui._('loginbutton')} aria-label={selfoss.ui._('loginbutton')} accesskey="l"></button>
+                <div class="nav-toolbar" id="nav_toolbar" aria-label={selfoss.ui._('nav_toolbar')} role="toolbar">
+                    <ul aria-labelledby="nav_toolbar">
```
Will not this be read twice? Would not it be sufficient to add `aria-label` and `role` here?

jtojnar replied:
Actually, does toolbar even make sense here? The buttons are grouped together mostly because there was not anywhere else to put them. Reading the [practices document](https://www.w3.org/TR/wai-aria-practices/#toolbar), it seems to me that it is mostly meant as an encapsulation for a large number of buttons, which is not really the case  here. Quite the contrary, users will want to be quickly able to access the settings page, not [rove](https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex) through a toolbar. The place where it makes sense to me is the entry action bar, where having to tab through all the sharing buttons while trying to get to the next entry would be annoying. We should implement roving behaviour there.

akash07k commented:
Actually the it was defined as a toolbar class so I thaught that
asigning the toolbar role to it would be much better.
Though, I've not implemented roving because there are not so many
buttons. I think toolbar role won't cause any problems for screen
reader      users. Should I remove the role?
But implementing the list for these buttons is good and in my opinion
it should not be removed.

On 10/12/19, Jan Tojnar <notifications@github.com> wrote:
> jtojnar commented on this pull request.
>
>
>
>> @@ -102,11 +102,13 @@ selfoss.ui = {
>                      <hr />
>                  </div>
>
> -                <div class="nav-toolbar">
> -                    <button id="nav-refresh"
> title={selfoss.ui._('refreshbutton')}
> aria-label={selfoss.ui._('refreshbutton')} accesskey="r"></button>
> -                    <button id="nav-settings"
> title={selfoss.ui._('settingsbutton')}
> aria-label={selfoss.ui._('settingsbutton')} accesskey="t"></button>
> -                    <button id="nav-logout"
> title={selfoss.ui._('logoutbutton')}
> aria-label={selfoss.ui._('logoutbutton')} accesskey="l"></button>
> -                    <button id="nav-login"
> title={selfoss.ui._('loginbutton')} aria-label={selfoss.ui._('loginbutton')}
> accesskey="l"></button>
> +                <div class="nav-toolbar" id="nav_toolbar"
> aria-label={selfoss.ui._('nav_toolbar')} role="toolbar">
> +                    <ul aria-labelledby="nav_toolbar">
>
> Actually, does toolbar even make sense here? The buttons are grouped
> together mostly because there was not anywhere else to put them. Reading the
> [practices document](https://www.w3.org/TR/wai-aria-practices/#toolbar), it
> seems to me that it is mostly meant as an encapsulation for a large number
> of buttons, which is not really the case  here. Quite the contrary, users
> will want to be quickly able to access the settings page, not
> [rove](https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex)
> through a toolbar. The place where it makes sense to me is the entry action
> bar, where having to tab through all the sharing buttons while trying to get
> to the next entry would be annoying. We should implement roving behaviour
> there.
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/SSilence/selfoss/pull/1151#discussion_r334207666

jtojnar commented:
> Actually the it was defined as a toolbar class

Yeah, I did not think about it much so I choose a class name based on how it looks, not what it was – a visual grouping, not a logical one.

> Though, I've not implemented roving because there are not so many buttons.

Yeah, I agree that it would be more annoying than useful here.

> I think toolbar role won't cause any problems for screen reader users.

How do screen readers react to a `toolbar` role? My main concern is the user will be told there is a toolbar and will try to use roving, which will fail, leaving them confused.

> Should I remove the role?

I would ask the following questions:

Does the role cause expectations that will not be fulfilled? If yes, can we change the code to meet those expectations? If no, then remove.

> But implementing the list for these buttons is good and in my opinion it should not be removed.

I do not mind the list but does it actually add anything? Do screen readers, for example, announce it as a list of buttons?

akash07k commented:
> > Actually the it was defined as a toolbar class
> 
> Yeah, I did not think about it much so I choose a class name based on how it looks, not what it was – a visual grouping, not a logical one.
> 
> > Though, I've not implemented roving because there are not so many buttons.
> 
> Yeah, I agree that it would be more annoying than useful here.
> 
> > I think toolbar role won't cause any problems for screen reader users.
> 
> How do screen readers react to a `toolbar` role? My main concern is the user will be told there is a toolbar and will try to use roving, which will fail, leaving them confused.
> 
> > Should I remove the role?
> 
> I would ask the following questions:
> 
> Does the role cause expectations that will not be fulfilled? If yes, can we change the code to meet those expectations? 
If no, then remove.
May be someone can think that since it is toolbar it will require roving navigation. I think then it should be removed altogether.
> 
> > But implementing the list for these buttons is good and in my opinion it should not be removed.
> 
> I do not mind the list but does it actually add anything? Do screen readers, for example, announce it as a list of buttons?
Yes, absolutely.

akash07k commented:
What should I do with it as of now? I'm unable to perform git rebase after pulling the latest changes from upstream, it seems that there are some heavy changes in selfoss-ui.js.

akash07k commented:
Tried to merge via `git rebase -X theirs` please check whether everything is fine or not. as per my investigation, everything seems right.

akash07k commented:
Still the checks are failing don't know why

jtojnar commented:
It appears it has not been rebased properly and there is some code that is not used any more. How are you rebasing?

You can get the failure explanation by clicking the Details link next to the check (Show all checks at the end of the page).

akash07k commented:
I'm rebasing it to master, Am pulling from the upstream remote into my
master and then rebasing the accessibility branch onto master. I'm
unable to get the error details either. :( don't know what's the
issue, seems that I'll have to disregard the whole branch again and
will have to start over from scratch

On 10/16/19, Jan Tojnar <notifications@github.com> wrote:
> It appears it has not been rebased properly and there is some code that is
> not used any more. How are you rebasing?
>
> You can get the failure explanation by clicking the Details link next to the
> check (Show all checks at the end of the page).
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/SSilence/selfoss/pull/1151#issuecomment-542800201

jtojnar commented:
Let me try to rebase it.

akash07k commented:
Ok bro. thanks in advance

On 10/16/19, Jan Tojnar <notifications@github.com> wrote:
> Let me try to rebase it.
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/SSilence/selfoss/pull/1151#issuecomment-542810086

jtojnar commented:
```shell-session
~/P/selfoss ((ce9d9538…)|…) $ git fetch git@github.com:akash07k/selfoss.git accessibility
From github.com:akash07k/selfoss
 * branch              accessibility -> FETCH_HEAD
~/P/selfoss ((ce9d9538…)|…) $ git checkout FETCH_HEAD
HEAD is now at ce9d9538 Merge branch 'accessibility' of https://github.com/akash07k/selfoss into accessibility
~/P/selfoss ((ce9d9538…)|…) $ git rebase SSilence/master
First, rewinding head to replay your work on top of it...
Applying: Added 'role="main"' to the login form since it is the only and main content at the page during login.
Applying: Fixed type for 'aria-labelledby' previously it was 'aria-labeledby'
Applying: Made the navigation toolbar more accessible:
Applying: Fixed the string for nav_toolbar string
Applying: Added the role=toolbar, it was missing previously
Applying: Added the 'assets/.cache' directory to the .gitignore list
Applying: Added 'role="main"' to the login form since it is the only and main content at the page during login.
Using index info to reconstruct a base tree...
M	assets/js/selfoss-ui.js
Falling back to patching base and 3-way merge...
Auto-merging assets/js/selfoss-ui.js
No changes -- Patch already applied.
Applying: Fixed type for 'aria-labelledby' previously it was 'aria-labeledby'
Using index info to reconstruct a base tree...
M	assets/js/selfoss-ui.js
Falling back to patching base and 3-way merge...
Auto-merging assets/js/selfoss-ui.js
No changes -- Patch already applied.
Applying: Made the navigation toolbar more accessible:
Using index info to reconstruct a base tree...
M	assets/js/selfoss-ui.js
M	assets/locale/en.json
Falling back to patching base and 3-way merge...
Auto-merging assets/locale/en.json
CONFLICT (content): Merge conflict in assets/locale/en.json
Auto-merging assets/js/selfoss-ui.js
CONFLICT (content): Merge conflict in assets/js/selfoss-ui.js
error: Failed to merge in the changes.
Patch failed at 0009 Made the navigation toolbar more accessible:
hint: Use 'git am --show-current-patch' to see the failed patch
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

Here I had to resolve the conflicts manually in the text editor.

```shell-session
~/P/selfoss (detached HEAD|REBASE 9/12|✖2✚2…) $ edit assets/locale/en.json
~/P/selfoss (detached HEAD|REBASE 9/12|✖2✚2…) $ git add assets/locale/en.json
~/P/selfoss (detached HEAD|REBASE 9/12|✖1✚1…) $ edit assets/js/selfoss-ui.js
~/P/selfoss (detached HEAD|REBASE 9/12|✖1✚1…) $ git add assets/js/selfoss-ui.js
~/P/selfoss (detached HEAD|REBASE 9/12|●1…) $ git rebase --continue
Applying: Made the navigation toolbar more accessible:
Applying: Fixed the string for nav_toolbar string
Using index info to reconstruct a base tree...
M	assets/locale/en.json
Falling back to patching base and 3-way merge...
Auto-merging assets/locale/en.json
No changes -- Patch already applied.
Applying: Added the role=toolbar, it was missing previously
Applying: Added the 'assets/.cache' directory to the .gitignore list
Using index info to reconstruct a base tree...
M	.gitignore
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
```

Now I rebased it few more times, manually fixing the incorrect changes from previous merges, adding necessary CSS, rewording the commit messages, and squashing commits that should be a single commit.

Finally I could push the changes back to your branch:

```ShellSession
~/P/selfoss ((328cd89e…)|…) $ git push git@github.com:akash07k/selfoss.git HEAD:accessibility -f
```

jtojnar commented:
I updated the comment above to make it more accessible. I wonder how can non-sighted users even use terminal when there are no landmarks, just a grid of text.

jtojnar commented:
You should be able to apply the changes locally by running `git fetch git@github.com:akash07k/selfoss.git accessibility` and `git reset --hard FETCH_HEAD`.

akash07k commented:
> This is how I did it:
> 
> 1. Obtained the fresh copy of your branch using `git fetch git@github.com:akash07k/selfoss.git accessibility`:
>    ```
>    From github.com:akash07k/selfoss
>     - branch              accessibility -> FETCH_HEAD
>    ```
> 2. Switched to the branch using `git checkout FETCH_HEAD`:
>    ```
>    HEAD is now at ce9d9538 Merge branch 'accessibility' of https://github.com/akash07k/selfoss into accessibility
>    ```
> 3. Rebased on the master using `git rebase SSilence/master`:
>    ```
>    First, rewinding head to replay your work on top of it...
>    Applying: Added 'role="main"' to the login form since it is the only and main content at the page during login.
>    Applying: Fixed type for 'aria-labelledby' previously it was 'aria-labeledby'
>    Applying: Made the navigation toolbar more accessible:
>    Applying: Fixed the string for nav_toolbar string
>    Applying: Added the role=toolbar, it was missing previously
>    Applying: Added the 'assets/.cache' directory to the .gitignore list
>    Applying: Added 'role="main"' to the login form since it is the only and main content at the page during login.
>    Using index info to reconstruct a base tree...
>    M   assets/js/selfoss-ui.js
>    Falling back to patching base and 3-way merge...
>    Auto-merging assets/js/selfoss-ui.js
>    No changes -- Patch already applied.
>    Applying: Fixed type for 'aria-labelledby' previously it was 'aria-labeledby'
>    Using index info to reconstruct a base tree...
>    M   assets/js/selfoss-ui.js
>    Falling back to patching base and 3-way merge...
>    Auto-merging assets/js/selfoss-ui.js
>    No changes -- Patch already applied.
>    Applying: Made the navigation toolbar more accessible:
>    Using index info to reconstruct a base tree...
>    M   assets/js/selfoss-ui.js
>    M   assets/locale/en.json
>    Falling back to patching base and 3-way merge...
>    Auto-merging assets/locale/en.json
>    CONFLICT (content): Merge conflict in assets/locale/en.json
>    Auto-merging assets/js/selfoss-ui.js
>    CONFLICT (content): Merge conflict in assets/js/selfoss-ui.js
>    error: Failed to merge in the changes.
>    Patch failed at 0009 Made the navigation toolbar more accessible:
>    hint: Use 'git am --show-current-patch' to see the failed patch
>    Resolve all conflicts manually, mark them as resolved with
>    "git add/rm <conflicted_files>", then run "git rebase --continue".
>    You can instead skip this commit: run "git rebase --skip".
>    To abort and get back to the state before "git rebase", run "git rebase --abort".
>    ```
> 4. Here I had to resolve the conflicts in `assets/locale/en.json` and `assets/js/selfoss-ui.js` manually in the text editor. Then I staged the changes using `git add assets/locale/en.json assets/js/selfoss-ui.js`
> 5. Now I could continue using `git rebase --continue`:
>    ```
>    Applying: Made the navigation toolbar more accessible:
>    Applying: Fixed the string for nav_toolbar string
>    Using index info to reconstruct a base tree...
>    M   assets/locale/en.json
>    Falling back to patching base and 3-way merge...
>    Auto-merging assets/locale/en.json
>    No changes -- Patch already applied.
>    Applying: Added the role=toolbar, it was missing previously
>    Applying: Added the 'assets/.cache' directory to the .gitignore list
>    Using index info to reconstruct a base tree...
>    M   .gitignore
>    Falling back to patching base and 3-way merge...
>    No changes -- Patch already applied.
>    ```
> 6. Now, I rebased it few more times, manually fixing the incorrect changes from previous merges, removing the `role=toolbar`, adding necessary CSS, rewording the commit messages, and squashing commits that should be a single commit.
> 7. Finally I could push the changes back to your branch `git push git@github.com:akash07k/selfoss.git HEAD:accessibility -f`

Oh, this was it. Actually I  also tried to resolve the conflicts manually but during this many things got broken because I'm not much efficient and experienced in performing git rebase. Also, many of the times terminal give a big pain with screen readers :(
Any ideas why it happened though my branch was clean?

akash07k commented:
> I updated the comment above to make it more accessible. I wonder how can non-sighted users even use terminal when there are no landmarks, just a grid of text.

Thanks a lot bro, :)
Actually we use terminals just like reading a text file in the note pad.
We have the commands to read line by line, character by character and so on.

But many times if terminal window contains a plenty of text we face problems while scrolling and diffing also becomes very difficult for use.
I have never been able to perform git diff successfully.

akash07k commented:
> You should be able to apply the changes locally by running `git fetch git@github.com:akash07k/selfoss.git accessibility` and `git reset --hard FETCH_HEAD`.

Applied the changes, and now getting this message:
Your branch and ’origin/accessibility’ have diverged, 
and have 4 and 13 different commits each, respectively. 
(use "git pull" to merge the remote branch into yours) 
Should I pull? I wonder why is it displaying if I've used git fetch.

akash07k commented:
> > You should be able to apply the changes locally by running `git fetch git@github.com:akash07k/selfoss.git accessibility` and `git reset --hard FETCH_HEAD`.
> 
> Applied the changes, and now getting this message:
> Your branch and ’origin/accessibility’ have diverged,
> and have 4 and 13 different commits each, respectively.
> (use "git pull" to merge the remote branch into yours)
> Should I pull? I wonder why is it displaying if I've used git fetch.

Oh well, I've pulled and now everything is up to date. :) Thanks a lot buddy.

jtojnar commented:
`git pull` on a diverged branch perform a merge, breaking the clean history we achieved using `rebase`. That’s why I have suggested to change the branch to point to the `FETCH_HEAD` (`origin/accessibility`) with the `git reset --hard FETCH_HEAD`

jtojnar commented:
Merged, thanks.

akash07k commented:
> `git pull` on a divergent branch perform a merge, breaking the clean history we achieved using `rebase`. That’s why I have suggested to change the branch to point to the `FETCH_HEAD` (`origin/accessibility`) with the `git reset --hard FETCH_HEAD`

Oh, I wasn't aware about it. Thanks bro. I'm learning a lot from you. :-)

akash07k commented:
> > But many times if terminal window contains a plenty of text we face problems while scrolling and diffing also becomes very difficult for use.
> 
> Yeah, I was afraid so. I was experimenting with a terminal, that would work with structured data instead of plain text (inspired by iTerm but much more ambitious). It would be more amenable to being accessible, but I did not have time to get it into a usable state. Maybe I will revisit it later.
Wow, great, so is it linux based terminal or cross platform?
> 
> > I have never been able to perform git diff successfully.
> 
> Yeah, `git diff` is tricky even for sighted users – one needs to install https://github.com/so-fancy/diff-so-fancy just to make it diff in a way that is useful.
> 
> You can actually [configure git](https://git-scm.com/docs/git-difftool) to use a different diff tool but I have no idea if there is any that is accessible.

Unfortunately as of now none of the difftool is properly accessible. Last year I was thinking to develop something by my own but later on dropped the idea due to the time investment and complexity of the work. :( I was very busy with my work those days.

akash07k commented:
> Merged, thanks.

:-)

jtojnar commented:
I am still not sure about the list in the sidebar navigation. For instance, the [accessible toolbar example](https://www.w3.org/TR/wai-aria-practices/examples/toolbar/toolbar.html) seems to work without them just fine. IMHO, we should add roving support to the filters/tags/sources, and everything in the sidebar, including the “toolbar” buttons will be considered a navigation item.

I have reverted f9861dc49bc6ef2d728a04cda754af1737509497 to facilitate that and also to make flex ox work correctly.

akash07k commented:
> I am still not sure about the list in the sidebar navigation. For instance, the [accessible toolbar example](https://www.w3.org/TR/wai-aria-practices/examples/toolbar/toolbar.html) seems to work without them just fine. IMHO, we should add roving support to the filters/tags/sources, and everything in the sidebar, including the “toolbar” buttons will be considered a navigation item.
> 
> I have reverted [f9861dc](https://github.com/SSilence/selfoss/commit/f9861dc49bc6ef2d728a04cda754af1737509497) to facilitate that and also to make flex ox work correctly.

I think that roving is not needed because there are not much buttons as of now. list is better. in fact if roving can be added then it is good. Though we are fine with it and as now the toolbar is removed I feel it is fine to only have the list of the buttons.
If we add the roving support to all the sidebar controls, then the navigation will become more easier and tab navigation will become relatively shorter. Though it will be very tricky to implement this because upon expanding the filters, tags etc, there will be many links as well.
